man/ctags-lang-python.7.rst.in

```diff
@@ -0,0 +1,167 @@
+::
+
+	X3	input.py	/^from X3 import Y3$/;"	kind:module	roles:namespace
+	Y3	input.py	/^from X3 import Y3$/;"	kind:unknown	scope:module:X3	roles:imported
+
+"Y3" is a name for a language object defined in "X3" module. "scope:module:X3"
+attached to "Y3" represents this relation between "Y3" and "X3". ctags
+assigns ``unknown`` kind to "Y3" because ctags cannot know whether "Y3" is a
+class, a variable, or a function from the input file.
+
+"input.py"
+.. code-block:: Python
+
+	from X4 import Y4 as Z4
+
+"output.tags"
+with "--options=NONE -o - --extras=+r --fields=+rzKZ input.py"
+::
+
+	X4	input.py	/^from X4 import Y4 as Z4$/;"	kind:module	roles:namespace
+	Y4	input.py	/^from X4 import Y4 as Z4$/;"	kind:unknown	scope:module:X4	roles:indirectlyImported
+	Z4	input.py	/^from X4 import Y4 as Z4$/;"	kind:unknown	roles:def	nameref:unknown:Y4
+
+"Y4" is similar to "Y3" of "from X3 import Y3" but the name cannot be used here.
+``indirectlyImported`` role assigned to "Y4" representing this. "Z4" is the name for
+accessing the language object named in "Y4" in "X4" module. "nameref:unknown:Y4"
+attached to "Z4" and "scope:module:X4" attached to "Y4" represent the relations.
+
+Lambda expression and type hint
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+"input.py"
+.. code-block:: Python
+
+	from typing import Callable
+	id = lambda var0: var0
+	id_t: Callable[[int], int] = lambda var1: var1
+
+"output.tags"
+with "--options=NONE -o - --sort=no --fields=+KS --fields-Python='+{nameref}' --extras='+{anonymous}' input.py"
+::
+
+	id	input.py	/^id = lambda var0: var0$/;"	function	signature:(var0)
+	id_t	input.py	/^id_t: Callable[[int], int] = lambda var1: var1$/;"\
+   		variable	typeref:typename:Callable[[int], int]	nameref:function:anonFunc84011d2c0101
+	anonFunc84011d2c0101	input.py	/^id_t: Callable[[int], int] = lambda var1: var1$/;"\
+   		function	signature:(var1)
+
+If a variable ("id") with no type hint is initialized with a lambda expression,
+ctags assigns ``function`` kind for the tag of "id".
+
+If a variable ("id_t") with a type hint is initialized with a lambda expression,
+ctags assigns ``variable`` kind for the tag of "id_t" with ``typeref:`` and
+``nameref:`` fields. ctags fills ``typeref:`` field with the value of the type
+hint. The way of filling ``nameref:`` is a bit complicated.
+
+For the lambda expression used in initializing the type-hint'ed variable, ctags
+creates ``anonymous`` extra tag ("anonFunc84011d2c0101"). ctags fills the
+``nameref:`` field of "id_t" with the name of ``anonymous`` extra tag:
+"nameref:function:anonFunc84011d2c0101".
+
+You may think why ctags does so complicated, and why ctags doesn't emit
+following tags output for the input::
+
+	id	input.py	/^id = \\$/;"	function	signature:(var0)
+	id_t	input.py	/^id_t: \\$/;"	function	typeref:typename:Callable[[int], int]	signature:(var1)
+
+There is a reason. The other languages of ctags obey following rule; ctags fills
```

Comment:
```suggestion
There is a reason. The other languages of ctags obey the following rule: ctags fills
```

Reply:
How the tag entry for `id_t` should be in your idea?

Reply:
I think it's good to just change `kind` to `function` and keep the rest. This is just my opinion. I'm not opposing tagging it as a variable, and I believe you have good reasons to do this.

Reply:
I have one extra idea to keep the kind of `id_t` function.

```
id_t: Callable[[int], int] = lambda var1: var1
```

If we can find "Callable" in the type hin for `id_t`, we can destruture it like:

```
id_t	input.py	/^id_t: \\$/;"	function	typeref:typename:int	signature:(var1: int)
```

This will not work well if Callable is not used directly.
How do you think about this idea?
I will implement this idea in the future if I get your support.

Reply:
This is ideal. However, I think if we want an accurate `kind` field for functions, we should do it for normal functions first, then deal with lambdas.

Reply:
I don't understand "for normal functions". Could you show me an example.
What I guessed:

```
def f ():
   pass

g = f
```
I guess what you say is "When tagging `g`, ctags should be able to assign "function" kind for `g`."

Am I correct?

Reply:
Oh, sorry for the typo. I mean "if we want an accurate `typeref` field". Here's an example:

``` python
def func(a: Foo) -> int:
    pass
```

For this, we should generate:

```
func ... typeref:typename:int
```

I think we should deal this before destructing the `Callable` type hint, since "lambdas with type hint" is much rarer than normal functions.

Reply:
The change, https://github.com/universal-ctags/ctags/pull/2613/commits/762a87fa2a6d9d4e3ca7538519e54cbac034096c
, makes ctags store the value on the right side of `->` to typeref field already.

What we have to do for normal function is just merging this pull request :-P.